### PR TITLE
Add tail and head to SeriesGroupby (#7934)

### LINKED ISSUE #7934: Series GroupBy head and tail
I'd like to be able to use head and tail on group-by Series. Today, I can do this in pandas but not dask.
```python
import pandas as pd
import dask.dataframe as dd

pdf = pd.DataFrame({
    "a": [0,1,0,1,1,0],
    "b": range(6),
    "c": ["a","b","c","d","e","f"]
})
ddf = dd.from_pandas(pdf, 2)

print(pdf.groupby("a").b.head(1))
ddf.groupby("a").b.head(1)
```
**Output**
```
0    0
1    1
Name: b, dtype: int64

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-1-6339dbe399f3> in <module>
     10 
     11 print(pdf.groupby("a").b.head(1))
---> 12 ddf.groupby("a").b.head(1)

AttributeError: 'SeriesGroupBy' object has no attribute 'head'
```

The same applies to `tail`. The pandas version is:
```python
dask.__version__
'2021.06.0'
```

### PR DESCRIPTION
**Pull request by mesejo**
- [x] Closes #7934 
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

### CODE REVIEW

File: dask/dataframe/groupby.py [hunk 1]
```diff
@@ -1972,3 +2000,33 @@ def _value_counts_aggregate(series_gb):
     to_concat = {k: v.groupby(level=1).sum() for k, v in series_gb}
     names = list(series_gb.obj.index.names)
     return pd.Series(pd.concat(to_concat, names=names))
+
+
+def _tail_meta(series_gb, **kwargs):
+    return series_gb.tail(**kwargs)
```
Comment (mrocklin) [inline dask/dataframe/groupby.py:2006]: This can probably be replaced with `M.tail`

File: dask/dataframe/groupby.py [hunk 2]
```diff
@@ -1972,3 +2000,33 @@ def _value_counts_aggregate(series_gb):
...
+        [group.tail(**kwargs) for group in groups], keys=keys, names=["_tail"]
+    )
+
+
+def _tail_aggregate(series_gb, **kwargs):
+    return series_gb.tail(**kwargs).droplevel("_tail")
+
+
+def _head_meta(series_gb, **kwargs):
+    return series_gb.head(**kwargs)
```
Comment (mrocklin) [inline dask/dataframe/groupby.py:2021]: This can probably be replaced with `M.head`.  This pattern is common enough that we made `M` to handle it.

File: dask/dataframe/tests/test_groupby.py [hunk 3]
```diff
@@ -611,6 +615,8 @@ def test_split_apply_combine_on_series(empty):
         assert_eq(ddf.groupby(ddf.b > i).b.size(), pdf.groupby(pdf.b > i).b.size())
         assert_eq(ddf.groupby(ddf.b > i).b.first(), pdf.groupby(pdf.b > i).b.first())
         assert_eq(ddf.groupby(ddf.b > i).b.last(), pdf.groupby(pdf.b > i).b.last())
+        assert_eq(ddf.groupby(ddf.b > i).b.tail(), pdf.groupby(pdf.b > i).b.tail())
+        assert_eq(ddf.groupby(ddf.b > i).b.head(), pdf.groupby(pdf.b > i).b.head())
```
Comment (mrocklin) [inline dask/dataframe/tests/test_groupby.py:619]: These tests are encouraging.  Thank you.

File: dask/dataframe/tests/test_groupby.py [hunk 4]
```diff
@@ -2520,3 +2528,91 @@ def test_groupby_with_pd_grouper():
...
+            ["a1", "b1"],
+            ["a1", "b1"],
+        ],
+        columns=["A", "B"],
+    )
+
+    expected = df.groupby("A")["B"].tail()
+    ddf = dd.from_pandas(df, npartitions=3)
+    actual = ddf.groupby("A")["B"].tail()
+    assert_eq(expected, actual)
```
Comment (mrocklin) [inline dask/dataframe/tests/test_groupby.py:2618]: Are there ways to combine any of these tests?  It might be nice to avoid a little bit of duplication here.  That is subjective though, feel free to ignore this comment.

File: dask/dataframe/groupby.py [hunk 5]
```diff
@@ -1143,6 +1143,7 @@ def _aca_agg(
         token,
         func,
         aggfunc=None,
+        metafunc=None,
```
Comment (mrocklin) [inline dask/dataframe/groupby.py:1146]: Making a change like this is concerning.  Why was this necessary?

Comment (mesejo) [reply dask/dataframe/groupby.py:1146]: If I use a function such as `M.head` for example then the grouping keys are not propagate to the `aggregate_func` (because [the pandas operation does not return the grouping keys](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.GroupBy.head.html)). Therefore I created a special function to be applied to the chunks: `_head_chunk`. This function returns a MultiIndex, if I use this function to compute the meta info then the index dtype wont match the pandas one. That is the reason that I decided to use two separate functions for computing the meta and applying to chunks. I could create a new function and leave `_aca_agg` unmodified if that is the concern. Any other other alternatives?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

File: dask/dataframe/tests/test_groupby.py (hunk 4 above)
Comment (jsignell) [reply dask/dataframe/tests/test_groupby.py:2497]: I assume you deduplicated, by parametrizing :ok_hand: - but it looks like you forgot to delete one of these tests. You have 2 tests called `test_groupby_with_row_operations`

File: dask/dataframe/groupby.py (hunk 5 above)
Comment (jsignell) [reply dask/dataframe/groupby.py:1143]: It would probably be more palatable to include an optional `meta` kwarg that accepts a precomputed `meta` rather than `metafunc`.

**Comment by quasiben** [comment]
add to allowlist

**Comment by mesejo** [comment]
> add to allowlist

@quasiben  Thanks for reviewing. I tried searching for allowlist but could not find it. Could you please expand on what needs to be done?

**Comment by mrocklin** [comment]
@quasiben this is exactly the kind of confusion I'd like to avoid with the GPU CI thing :)

@mesejo you're fine.  My apologies that you haven't gotten much review here.  I suspect that what happened is that because people saw me engage on this PR other reviewers have assumed that I'm taking care of it, when in fact I've been distracted by other things.  People like @jrbourbeau and @jsignell often review PRs like this one.  Hopefully one of them will be by in the next couple of days.

File: dask/dataframe/groupby.py (hunk 5 above)
Comment (mesejo) [reply dask/dataframe/groupby.py]: @jsignell Yes, you are right.

**Comment by mrocklin** [comment]
This looks good to me.  Thank you for your work and your patience @mesejo !

**Comment by mesejo** [comment]
> This looks good to me. Thank you for your work and your patience @mesejo !

No problem @mrocklin. Glad I could help!
